src/ffi.rs

```diff
@@ -40,3 +44,271 @@ pub fn disable_internal_nscd() {
         __nss_disable_nscd(do_nothing);
     }
 }
+
+pub enum LibcIp {
+    V4([u8; 4]),
+    V6([u8; 16])
+}
+
+mod glibcffi {
+    use nix::libc;
+    extern "C" {
+        pub fn gethostbyname2_r (
+            name: *const libc::c_char,
+            af: libc::c_int,
+            result_buf: *mut libc::hostent,
+            buf: *mut libc::c_char,
+            buflen: libc::size_t,
+            result: *mut *mut libc::hostent,
+            h_errnop: *mut libc::c_int,
+        ) -> libc::c_int;
+
+        pub fn gethostbyaddr_r (
+            addr: *const libc::c_void,
+            len: libc::socklen_t,
+            af: libc::c_int,
+            ret: *mut libc::hostent,
+            buf: *mut libc::c_char,
+            buflen: libc::size_t,
+            result: *mut *mut libc::hostent,
+            h_errnop: *mut libc::c_int,
+        ) -> libc::c_int;
+    }
+}
+
+/// This structure is the Rust counterpart of the `libc::hostent` C
+/// function the Libc hostent struct.
+///
+/// It's mostly used to perform the gethostbyaddr and gethostbyname
+/// operations.
+///
+/// This struct can be serialized to the wire through the
+/// `serialize` function or retrieved from the C boundary using the
+/// TryFrom `libc:hostent` trait.
+#[derive(Clone, Debug)]
+pub struct Hostent {
+    pub name: String,
+    pub aliases: Vec<String>,
+    pub addr_type: i32,
+    pub addr_list: Vec<std::net::IpAddr>,
+}
+
+impl TryFrom<libc::hostent> for Hostent {
+    type Error = anyhow::Error;
+
+    fn try_from(value: libc::hostent) -> Result<Self, Self::Error> {
+        // validate value.h_addtype, and bail out if it's unsupported
+        if value.h_addrtype != libc::AF_INET && value.h_addrtype != libc::AF_INET6 {
+            bail!("unsupported address type: {}", value.h_addrtype);
+        }
+
+        // ensure value.h_length matches what we know from this address family
+        if value.h_addrtype == libc::AF_INET && value.h_length != 4 {
+            bail!("unsupported h_length for AF_INET: {}", value.h_length);
+        }
+        if value.h_addrtype == libc::AF_INET6 && value.h_length != 16 {
+            bail!("unsupported h_length for AF_INET6: {}", value.h_length);
+        }
+
+        let name = unsafe { CStr::from_ptr(value.h_name).to_str().unwrap().to_string() };
+
+        Ok({
+            // construct the list of aliases. keep adding to value.h_aliases until we encounter a null pointer.
+            let mut aliases: Vec<String> = Vec::new();
+            let mut h_alias_ptr = value.h_aliases;
+            loop {
+                if unsafe {(*h_alias_ptr).is_null()} {
```

Comment:
I don't think you don't want to dereference h_alias_ptr here but instead check if h_alias_ptr itself is null.

Reply:
Also you can use the break condition to a while loop.